internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,666 @@
+                    v: pipe.v.clone(),
+                });
+            }
+        }
+
+        args.append(&mut req);
+        args.append(&mut opt);
+        match &self.retn {
+            Some(t) => write!(f, "({}) -> {}", List { values: &args }, t),
+            None => write!(f, "({}) -> !", List { values: &args }),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct BoundTvar {
+    tv: Tvar,
+    kind: Kind,
+}
+
+impl fmt::Display for BoundTvar {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}:{}", self.tv, self.kind)
+    }
+}
+
+// List is a wrapper around a generic vector
+#[derive(Debug, Clone, PartialEq)]
+struct List<'a, T> {
+    values: &'a Vec<T>,
+}
+
+// Display a generic list whose values can be displayed
+impl<'a, T: fmt::Display> fmt::Display for List<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.values.is_empty() {
+            return Ok(());
+        }
+        let size = self.values.len();
+        let list = &self.values[..size - 1];
+        for v in list {
+            if let Err(e) = write!(f, "{}, ", v) {
+                return Err(e);
+            }
+        }
+        self.values[size - 1].fmt(f)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn display_kind_add() {
+        assert!(Kind::Add.to_string() == "Addable");
+    }
+    #[test]
+    fn display_kind_div() {
+        assert!(Kind::Div.to_string() == "Divisible");
+    }
+    #[test]
+    fn display_kind_cmp() {
+        assert!(Kind::Cmp.to_string() == "Comparable");
+    }
+    #[test]
+    fn display_kind_nul() {
+        assert!(Kind::Nul.to_string() == "Nullable");
+    }
+
+    #[test]
+    fn display_type_bool() {
+        assert_eq!("bool", MonoType::Bool.to_string());
+    }
+    #[test]
+    fn display_type_int() {
+        assert_eq!("int", MonoType::Int.to_string());
+    }
+    #[test]
+    fn display_type_uint() {
+        assert_eq!("uint", MonoType::Uint.to_string());
+    }
+    #[test]
+    fn display_type_flt() {
+        assert_eq!("flt", MonoType::Flt.to_string());
+    }
+    #[test]
+    fn display_type_str() {
+        assert_eq!("str", MonoType::Str.to_string());
+    }
+    #[test]
+    fn display_type_dur() {
+        assert_eq!("dur", MonoType::Dur.to_string());
+    }
+    #[test]
+    fn display_type_time() {
+        assert_eq!("time", MonoType::Time.to_string());
+    }
+    #[test]
+    fn display_type_rgx() {
+        assert_eq!("rgx", MonoType::Rgx.to_string());
+    }
+    #[test]
+    fn display_type_tvar() {
+        assert_eq!("t10", MonoType::Var(Tvar(10)).to_string());
+    }
+    #[test]
+    fn display_type_array() {
+        assert_eq!(
+            "[int]",
+            MonoType::Arr(Box::new(Array(MonoType::Int))).to_string()
+        );
+    }
+    #[test]
+    fn display_type_record() {
+        assert_eq!(
+            "{a:int | {b:str | t0}}",
```

Comment:
Do we want to flatten these for display?

Reply:
Yes we do.